### ISSUE TITLE
Associate costing options with country

### ISSUE THREAD

**Issue by ADMUKHE** [valhalla/valhalla#4161]
Currently, Valhalla lacks a functionality that would enable costing options to be connected to particular countries. 

We are unable to accomodate any special request where one might want to travel from **Point A(in Germany) to Point B(in Switzerland)** and would want to avoid tolls in Germany and  highways in Switzerland. The route request would then look like:

```
{
"locations": [
    {
      "lon": -122.990975,
      "lat": 42.305183
    },
    {
      "lon": -123.048806,
      "lat": 42.277015
    }
  ],
  "costing": "auto",
  "costing_options": {
    "auto": {
      {"use_highways": 0, "country_code": ["XXX"] },
      {"use_tolls": 0, "country_code": ["XXX", "XXA"] }
    }
  }
} 
 ```

Instead of a single country_code, we can have a list as well. An option where no country_code is mentioned, by default it is valid for all.

Is it feasible to include such a functionality in the repository? 

Can the Route API be extended (without causing any intermediate components to break)? Given how complex it can become(connecting the option with country code, modifying the request parser and so on), perhaps a list of the components that need to be addressed might make more sense.

**Comment by nilsnolde** [comment]
Yeah that’d be nice to have, and similar to #4142. Having a generic interface for country dependent costing options would be even better, but I can’t see how that’ll be possible in the current state. It’ll need to be backwards compatible while not grossly uglifying the code base. As @kevinkreiser mentioned before somewhere, it likely would.

However, if someone dives in and presents a reasonable (API & code) design, I’m sure we could agree on smth.

**Comment by nilsnolde** [comment]
Yeah that’d be nice to have, and similar to #4142. Having a generic interface for country dependent costing options would be even better, but I can’t see how that’ll be possible in the current state. It’ll need to be backwards compatible while not grossly uglifying the code base. As @kevinkreiser mentioned before somewhere, it likely would.

However, if someone dives in and presents a reasonable (API & code) design, I’m sure we could agree on smth.

**Comment by ADMUKHE** [comment]
What structure would a request schema for country dependent costing option look like? Do you have any idea in mind?

I was wondering, which alternative would be better suited as a general feature and why! Costing option per country or associating country codes to avoid options. What are the pros of having costing option per country when associating avoid options with specific country can acheive the same goal!

**Comment by ADMUKHE** [comment]
I am not able to wrap my head around this "Having a generic interface for country dependent costing options would be even better". what do you mean by country dependent costing options? something like a set of options associated with a list of countries or each costing option being assigned to a value and a list of countries like  
``` { "use_tolls" : [0, "DE", "CH"], use_highways: 1 }  ```

**Comment by nilsnolde** [comment]
> what do you mean by country dependent costing options

I meant your suggestion :smile: as opposed to #4142 which only talks about a single costing option.

I realize it's a bit too much asked of someone who's not familiar with Valhalla. For me/us it's very hard to give suggestions without having the time to sit down and sketch things out properly.

We definitely can't introduce more nesting as you suggest here: https://github.com/valhalla/valhalla/issues/4161#issuecomment-1586135292. And this one is not nice as an interface: https://github.com/valhalla/valhalla/issues/4161#issuecomment-1588684442.

Honestly, I feel this is more v4 material with a completely overhauled API and we can't accommodate it in the current API. I might be wrong, but I don't want to waste your time either..

**Comment by nilsnolde** [comment]
FWIW, I'd be mostly in favor of appending the country code to the costing model, e.g. `"costing_options": {"auto_de_fr"}: {...}` or just having another member `country_codes` inside the `"auto"` costing options. It's not perfect either, but would be backwards-compatible. 

But that's just the interface. The code is a completely different topic..

**Comment by kevinkreiser** [comment]
assuming we had the api design done (the bulk of this discussion so far has been about that)  enabling this functionality within the costing is the other thing to tackle. @ADMUKHE made another issue (#4194) for this but ive closed it in favor of answering it here and keeping the discussion in one place.

to me the solution for making this possible would somehow modifying the transition cost to penalize switching to a new admin. lets think this out.

we currently have a border penalty and many issues that complain that it does nothing. so clearly even the blanket avoid any country crossing doesnt work. why is that... well thats because we do it based on node type: https://github.com/valhalla/valhalla/blob/26489fbc614db2b0d33c8f38ea1490054905a7c5/valhalla/sif/dynamiccost.h#L1120C5-L1120C85

and we only ever set that if the damned thing is actually tagged that way in osm: https://github.com/valhalla/valhalla/blob/26489fbc614db2b0d33c8f38ea1490054905a7c5/src/mjolnir/pbfgraphparser.cc#L1618

oops i guess "country crossing" is kind of a misnomer. anyway i digress... a redesign of this feature to do what you actually expect should probably actually check the admins of the nodes in the graph as @nilsnolde alluded to.

for me, this means that the transition cost, ie the time at which you evaluate whether or not you are going to enter a new admin should be the place where we penalize the crossing whether specific or generic. at first glance it seems kind of possible...

transition cost has the predecessor edge information and the current node. we could attach admin information to the predecessor edge label and compare it to the admin info on the current node and if they differ we penalize.

sounds easy right? well not really. attaching admin info to the predecessor edge is hard. for one thing, the admin information for a given node is local to its tile. it stores an index into a list of all the admins that exist in a specific tile. a tile that borders the us and canada has at least two admins in it but an adjacent tile might only have one in the us. since the canada could have been index 0 in the first tile but us might be at index 0 in the second you cant just compare these indices. so thats the first problem.

the second problem is even if you could compare the indices we dont know, from the index that which country it is, which is the thrust of this issue. we need to know which country it is to do country specific logic.

however the news is not all bad! the actual admin records in the tile come with their 2 char iso codes directly attached to them. what this means is that we can get that 2 char country code with a very quick lookup so long as we have the tile.

thats the next problem, transition cost doesnt get the tile. but heck maybe it should? we have the node which means the caller also has the tile. surely we can pass it on? its only fair, edgecost also takes the tile. to me this only makes sense, and would make the transition cost more of a first class citizen like edgecost is. still its a lot of change. and still it doesnt fix all the problems.

remember that the predecessor edge could be in a different tile than the end node. that is why to fully implement this the costing would need to cache the admin of every unique `tile_id, nodeinfo.admin_index` pair that it sees. if yuou do this from the beginning of the expansion you are gauranteed never to have a cache miss when looking up your predecessor edges admin info (in the case the predecessor edge is in a different, but importantly previously seen and cached, tile).

ok so that was a long description but, long story short, its possible for costing to know every time you get to a node whether or not you are changing admins and from which ones. we'd just need to:

1. update dynamic_cost::transition_cost to take a graph tile
2. update dynamic_cost::transition_cost to cache the admin entry for each node it sees (cheap it will be a small number in almost all cases but we can reserve as we see fit)
3. update edgelabel to track the admin index of its begin node (or an invalid one for the first edge label in an expansion)

after we did that we could properly implement generic country crossing penalties (based on admin area polygons not border control points, maybe the existing api naming is just a bad choice) or we could get more complicated and implement some country specific stuff.

**Comment by ADMUKHE** [comment]
The goal is to work on a specific use case where we want to do something like "use_tolls_per_country" = ["DE", "CH" ]. Checking if the edge is toll and **adding a penalty for using edge that is a toll and belong to a specific country** in the **transition cost** seems acheivable from your explaination. 

However, why do you think we cannot do this in the EdgeCost(is it only because of the missing info like the reader to get the node info )? Or adding a penalty for specific conditions like avoid tolls specific to country in EdgeCost would not produce the desired results and are a better fit for the transitional cost?

**Comment by kevinkreiser** [comment]
@ADMUKHE i was taking your several related issues and trying to put all the information in one place. in doing so i focused on what i thought was the most difficult (to implement) one, which is specific country to specific other country transitions. this would need to happen in transition_cost and having to deal with 2 at once across tiles makes it hard. for the easier country specific costing options sometimes the right place to add the penalty would be in edge_cost, for the toll example that is definitely the case.

basically i was trying to design a general solution that, if we implemented it, would open up all the possibilities you seem to be interested in rather than focus on the specific one in this issue.

honestly i find that multiple related issues (we have already close to 600 total?) is really cluttering up the repository. its better to group like things together not only for the ease of searching through them but also to come up with general rather than specific solutions to larger overall problems. we've failed at this mostly i'd say.

anyway to the specific issue yes edgecost is the right place for a penalty that scales with the length of the portion of the path that incurs the penalty. how we would implement it would be to do everything i said above, keep each country specific costing object in a wrapper object (maybe the admin cache too that ive been mentioning). then each algorithm has to pull the right costing to use based on the predecessor before calling any transition/edgecost functions on it.

**Comment by ADMUKHE** [comment]
Are you suggesting to do the same in the edge_cost like update with an admin cache? we also can have the country info from reader, tile and de. Why not pass the reader in the edgecost from the algorithms? 

yes, your approach seems better for our final target to combine costing specific to countries.

**Comment by kevinkreiser** [comment]
costing is called in the tight loop of every algorithm. passing the reader would allow ultimate control yes but it would also allow a person to do graph traversals inside of the costing functions. where possible we should dissuade people from making such mistakes. Basically we can make sure costing stays performant by not allowing it do to too many complicated things. if costing has access to the whole graph we throw that protection out the window. not a good idea in my opinion

**Comment by xxhh1112** [comment]
t
